src/geometry/manifold/algebra/left_invariant_derivation.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.derivation_bundle
+import ring_theory.derivation
+
+/-!
+
+# Left invariant derivations
+
+In this file we define the concept of left invariant derivation for a Lie group. The concept is
+analogous to the more classical concept of left invariant vector fields, and it holds that the
+derivation associated to a vector field is left invariant iff the the field is.
+
+Moreover we prove that `left_invariant_derivation I G` has the structure of a Lie algebra, hence
+implementing one of the possible definitions of Lie algebra.
+
+-/
+
+noncomputable theory
+
+open_locale lie_group manifold
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+(G : Type*) [topological_space G] [charted_space H G] [monoid G] [has_smooth_mul I G] (g h : G)
+
+-- Generate trivial has_sizeof instance. It prevents weird type class inference timeout problems
+local attribute [instance, priority 10000]
+private def disable_has_sizeof {α} : has_sizeof α := ⟨λ _, 0⟩
+
+/--
+Left-invariant global derivations.
+
+A global derivation is left-invariant if it is equal to its pullback along left multiplication by
+an arbitrary element of `G`.
+-/
+structure left_invariant_derivation extends derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯ :=
+(left_invariant'' : ∀ f g, 𝒅(𝑳 I g) 1 (derivation.eval_at 1 to_derivation) f =
+  derivation.eval_at g to_derivation f)
+
+variables {I G}
+
+namespace left_invariant_derivation
+
+instance : has_coe (left_invariant_derivation I G) (derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯)
+:= ⟨λ X, X.to_derivation⟩
+
+instance : has_coe_to_fun (left_invariant_derivation I G) := ⟨_, λ X, X.to_derivation.to_fun⟩
+
+variables
+{M : Type*} [topological_space M] [charted_space H M] {x : M}
+{X Y : left_invariant_derivation I G} {f f' : C^∞⟮I, G; 𝕜⟯} {r : 𝕜}
+
+lemma to_fun_eq_coe : X.to_fun = ⇑X := rfl
+lemma coe_fn_coe : ⇑(X : C^∞⟮I, G; 𝕜⟯ →ₗ[𝕜] C^∞⟮I, G; 𝕜⟯) = X := rfl
+@[simp] lemma to_derivation_eq_coe : X.to_derivation = X := rfl
+
+lemma coe_injective (h : ⇑X = Y) : X = Y :=
```

Comment:
This gives a little more API:
```suggestion
lemma coe_injective : @function.injective (left_invariant_derivation I G) _ coe_fn :=
```
I have no idea what type should fil in that underscore, but hopefully you do - it's the same as the underscore on line 53

Reply:
What kind of additional API does this give?

Reply:
I mean isn't this lemma standard in this form all across mathlib?

Reply:
No, for instance we have [`equiv.coe_fn_injective`](https://leanprover-community.github.io/mathlib_docs/data/equiv/basic.html#equiv.coe_fn_injective) stated in the way I suggest. The version stated the way you have it here is usually called `coe_inj`; `injective` is reserved by the naming guide for things about `function.injective`.

Reply:
```suggestion
lemma coe_injective :
  @function.injective (left_invariant_derivation I G) (_ → C^⊤⟮I, G; 𝕜⟯) coe_fn :=
λ X Y h, by { cases X, cases Y, congr', exact derivation.coe_injective h }

```

Reply:
Ok done!